Allow gPTP over VLAN

jukkar opened the pull request.

jukkar committed:
samples: net: gptp: Allow running gPTP over VLAN

By default gPTP is not run over VLAN but if needed that can be
done by setting CONFIG_NET_GPTP_VLAN and CONFIG_NET_GPTP_VLAN_TAG
options.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

tbursztyka commented on subsys/net/l2/ethernet/gptp/gptp_messages.c:
```diff
@@ -129,10 +129,25 @@ struct net_pkt *gptp_prepare_sync(int port)
 	NET_ASSERT(iface);
 
+#if defined(CONFIG_NET_VLAN)
+	eth_ctx = net_if_l2_data(iface);
```
shouldn't you be using net_eth_is_vlan_enabled() here?

Btw, I can't recall why net_eth_is_vlan_enabled() has to take the context in parameter? Since we pass the iface, it can get the context. Or is this function meant to be used with a context that does not belong to the iface? (which sounds a bit weird)

jukkar replied:
I will introduce a small helper to set the vlan header.

tbursztyka commented on subsys/net/l2/ethernet/gptp/gptp_messages.c:
```diff
@@ -129,10 +129,25 @@ struct net_pkt *gptp_prepare_sync(int port)
 	struct gptp_hdr *hdr;
 
+#if defined(CONFIG_NET_VLAN)
+	struct net_eth_vlan_hdr *hdr_vlan;
```
move that after line 202

jukkar replied:
Hmm, I think that is against the coding style. Meaning we should introduce variables at the beginning of the function.

jukkar replied:
This cannot be done because hdr_vlan is filled later in the function. For that we also need the vlan_enabled flag.

tbursztyka commented on subsys/net/l2/ethernet/gptp/gptp_messages.c:
```diff
@@ -149,6 +164,12 @@ struct net_pkt *gptp_prepare_sync(int port)
 
+#if defined(CONFIG_NET_VLAN)
+	if (vlan_enabled) {
+		hdr_vlan = (struct net_eth_vlan_hdr *)NET_ETH_HDR(pkt);
```
move that after line 202

tbursztyka commented on subsys/net/l2/ethernet/gptp/gptp_messages.c:
```diff
@@ -212,11 +242,26 @@ struct net_pkt *gptp_prepare_follow_up(int port, struct net_pkt *sync)
 	struct net_buf *frag;
 
+#if defined(CONFIG_NET_VLAN)
+	struct net_eth_vlan_hdr *hdr_vlan;
```
same, move after line 317

tbursztyka commented on subsys/net/l2/ethernet/gptp/gptp_messages.c:
```diff
@@ -232,6 +277,13 @@ struct net_pkt *gptp_prepare_follow_up(int port, struct net_pkt *sync)
+#if defined(CONFIG_NET_VLAN)
+	if (vlan_enabled) {
+		hdr_vlan = (struct net_eth_vlan_hdr *)NET_ETH_HDR(pkt);
+		net_pkt_set_vlan_tag(pkt, net_pkt_vlan_tag(sync));
```
move these 2 lines after 317

tbursztyka commented on subsys/net/l2/ethernet/gptp/gptp_messages.c:
```diff
@@ -294,10 +355,25 @@ struct net_pkt *gptp_prepare_pdelay_req(int port)
 	struct gptp_hdr *hdr;
 
+#if defined(CONFIG_NET_VLAN)
+	struct net_eth_vlan_hdr *hdr_vlan;
```
well you know the song now.

tbursztyka reviewed (changes requested):
Why can't this eth header filling be factorized? it's the same exact code copy pasted all over.

well in short having a higher level protocol messing up with link layer is a super mess. It's what we've been doing so far and that's going to disappear (it has to, better have a single place for filling in l2 header than scattered everywhere).

jukkar committed:
net: eth: Add helper to return VLAN info for an interface

A small helper function will return information whether
a given network interface has VLAN enabled or not.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: mgmt: Add VLAN enabled / disabled event support

Send network management event if VLAN tag is enabled or disabled.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: gptp: Allow gPTP to run over VLAN

Allow this setup as Linux supports this too.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

tbursztyka commented on subsys/net/l2/ethernet/gptp/gptp_messages.c:
```diff
@@ -178,7 +209,11 @@ struct net_pkt *gptp_prepare_sync(int port)
 	hdr->reserved2 = 0;
 
 	/* Ethernet configuration. */
+#if defined(CONFIG_NET_GPTP_VLAN)
```
here it would be:
```
if (IS_ENABLED(CONFIG_NET_GPTP_VLAN)) {
        setup_eth_vlan_hdr(iface, pkt);
} else {
        eth->type = htons(NET_ETH_PTYPE_PTP);
}
```

tbursztyka commented on subsys/net/l2/ethernet/gptp/gptp_messages.c:
```diff
@@ -129,10 +145,19 @@ struct net_pkt *gptp_prepare_sync(int port)
+#if defined(CONFIG_NET_GPTP_VLAN)
+	vlan_enabled = net_eth_get_vlan_info(iface, &eth_len);
+#endif
+
```
line 166:
```
frag = gptp_get_frag() and make that function, aware of vlan so reserving the right amount etc...
Then no need of eth_len being set here if vlan or not. This logic is repeated all over below, so it's worth getting 1 place doing it.

}
```

jukkar commented on subsys/net/l2/ethernet/ethernet_mgmt.c:
```diff
@@ -121,3 +121,24 @@ void ethernet_mgmt_raise_carrier_off_event(struct net_if *iface)
+void ethernet_mgmt_raise_vlan_enabled_event(struct net_if *iface, u16_t tag)
+{
+#if defined(CONFIG_NET_MGMT_EVENT_INFO)
+	net_mgmt_event_notify_with_info(NET_EVENT_ETHERNET_VLAN_TAG_ENABLED,
```
For enabled event we could get it from iface, but that would not work for disabled event. Thus in order to keep the functions looking similar, I opted to have the tag in the enabled API.

jukkar commented:
More refactoring done

jukkar commented on subsys/net/l2/ethernet/ethernet.c:
```diff
@@ -651,6 +651,22 @@ u16_t net_eth_get_vlan_tag(struct net_if *iface)
 	return NET_VLAN_TAG_UNSPEC;
 }
 
+bool net_eth_get_vlan_info(struct net_if *iface, int *eth_len)
```
Lets get rid of that second parameter, it is not really needed.

tbursztyka reviewed (approved):
Yeah now it's just right
